spec/lib/openc_bot/helpers/register_methods_spec.rb

```diff
@@ -176,12 +176,13 @@ module ModuleWithNoCustomPrimaryKey
       ModuleThatIncludesRegisterMethods.save_data([:custom_uid], custom_uid: "A094567")
     end
 
-    it "gets entries which have not been retrieved or are more than 1 month old, oldest first" do
+    it "gets entries which have not been retrieved or are more than 1 month old, active first followed by inactive (with duplicates unless we process them), and oldest first within each set" do
       ModuleThatIncludesRegisterMethods.save_data([:custom_uid], custom_uid: "5234888", retrieved_at: (Date.today - 40).to_time)
+      ModuleThatIncludesRegisterMethods.save_data([:custom_uid], custom_uid: "3341431", retrieved_at: (Date.today - 45).to_time, dissolution_date: Date.today) # inactive
       ModuleThatIncludesRegisterMethods.save_data([:custom_uid], custom_uid: "87654", retrieved_at: (Date.today - 50)) # date, not time.
       ModuleThatIncludesRegisterMethods.save_data([:custom_uid], custom_uid: "9234567", retrieved_at: (Date.today - 2).to_time) # not stale
 
-      expect { |b| ModuleThatIncludesRegisterMethods.stale_entry_uids(&b) }.to yield_successive_args("99999", "A094567", "87654", "5234888")
+      expect { |b| ModuleThatIncludesRegisterMethods.stale_entry_uids(&b) }.to yield_successive_args("99999", "A094567", "87654", "5234888", "99999", "A094567", "87654", "3341431", "5234888")
```

Comment:
This test shows that the method can yield a lot of duplication as the 2nd query retrieves the same records that the 1st query did (plus more because it also picks up on inactives)
...however this only happens in the test and not in reality, because in reality we're doing _updates_ on the yielded records, so when we come to run the 2nd query it won't pick up on the same records (Updated records get a recent retrieved_at)

Foolishly I tried to fix this by adding a clause into the second query:
`WHERE ocdata.#{primary_key_name}  NOT IN (#{active_sql}) AND ...`

I got all that working, and it removed the duplicates nicely. But just now I realised I'd created the opposite problem. Now it's re-running the 1st query and when running in reality that will mean it grabs a _new_ bunch of active ids and removes them from the results of the second query!

...so I put it back like this again. It means the method assumes that we are _processing_ the yielded items, or at least it can yield duplicates if we are not. A duplicate in may not be a big problem

Reply:
I've resolved this in a fairly brute force way. Keep an array of all the active company ids we've processes from the 1st query, and skip them if we get them again in the 2nd.

It's going to throw the LIMIT count a bit if we get lots of these "skip" records in the 2nd query. Our calculated LIMIT on the 2nd query doesn't take account of records we'll be skipping over, which will mean we end up processing less than the desired stale_count. However it only happens on failures. Worst case scenario is: 0.5 ratio. All the 1st query fail. All the 2nd query skip. So we only "process" half the desired count. But with that many failures happening, the fact that we didn't process as many as we would've liked will be the least of our worries (bot needs fixing anyway!)